prov/shm/src/smr_hmem.h

```diff
@@ -0,0 +1,74 @@
+
+/*
+ * Copyright (c) Amazon.com, Inc. or its affiliates. All rights reserved.
+ * Copyright (c) Intel Corporation. All rights reserved
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _HMEM_SHM_H_
+#define _HMEM_SHM_H_
+
+#include "ofi_hmem.h"
+#include "smr.h"
+
+static inline ssize_t smr_copy_from_mr_iov(void *dest, size_t size,
+                                           struct ofi_mr **mr,
+                                           const struct iovec *iov,
+                                           size_t iov_count,
+                                           uint64_t iov_offset)
+{
+    if (FI_VERSION_GE(smr_prov.fi_version, FI_VERSION(1,19)) &&
+        1 == iov_count && mr && mr[0] && FI_HMEM_CUDA == mr[0]->iface &&
```

Comment:
Why `1 == iov_count`?

Reply:
i think it is following the ompi style.

Reply:
I remember there was such condition in shm for ipc, I guess here is that we may not want to support mixed iface type.

Reply:
Yes. SHM currently does not support mixed HMEM iface. Here is an example: https://github.com/ofiwg/libfabric/blob/main/prov/shm/src/smr_msg.c#L312

Reply:
shm does not support mixed iface for the IPC path but does support mixed ifaces in the inline/inject paths which is what this is optimizing. I added the ofi_copy_from_mr_iov functions for that reason.
I'm thinking it could be better to have the gdr copy size limit be global and move this to the mr_iov copy path since the hmem data is part of the mr info. thoughts?

Reply:
> shm does not support mixed iface for the IPC path but does support mixed ifaces in the inline/inject paths which is what this is optimizing.

Actually we are optimizing the current cuda ipc path, which only supports 1 hmem iface as mentioned. But what you are suggesting here also makes sense:

> it could be better to have the gdr copy size limit be global ...

It could be simpler to make the size limit global (because this limit is only related to the hardware and cuda driver, not efa/shm/etc provider). 

>  ... and move this to the mr_iov copy path

I considered this option but it sorta conflicted with our intention of separating gdrcopy from generic hmem copy.  Our idea is to make the caller explicitly choose the copy method. Therefore hiding gdrcopy inside mr iov copy path sounds contradictory to this end.

You pointed out a valid issue - the current impl will not use gdrcopy for inline/inject paths **when** there are mixed ifaces. This is not an issue here since we are optimizing the cuda ipc path that has only 1 iface, i.e.
```
	/* Do not inline/inject if IPC is available so device to device
	 * transfer may occur if possible. */
	if (iov_count == 1 && desc && desc[0]) {
		smr_desc = (struct ofi_mr *) *desc;
		iface = smr_desc->iface;
		use_ipc = ofi_hmem_is_ipc_enabled(iface) &&
				smr_desc->flags & FI_HMEM_DEVICE_ONLY &&
				!(op_flags & FI_INJECT);

		if (FI_VERSION_GE(smr_prov.fi_version, FI_VERSION(1,19)) &&
		    FI_HMEM_CUDA == iface &&
		    (OFI_HMEM_DATA_GDRCOPY_HANDLE & smr_desc->flags)) {
			assert(smr_desc->hmem_data);
			gdrcopy_available = true;
		}
	}
```

An alternative to changing the common mr iov path is to change e.g. `smr_copy_from_mr_iov` and iterate over the ifaces, then we can choose the proper memcopy method case by case. The downside is code duplication with mr iov copy, plus we don't have an immediate use for that.

Reply:
I'm a little confused by this. You're saying you're trying to target the IPC path for the gdrcopy optimization but the previous PR that added the gdrcopy size limit was directing it to the inline and inject path, not IPC. And in this patch you're updating all the copy functions (including the inline and inject paths) to use gdrcopy if it's available. What is the ultimate goal for these optimizations?

Reply:
@aingerson Yes it sounds a little confusing. What I meant is that, before this change, cuda IPC would have been selected for cuda->cuda data transfers via cudaMemcpy. After this change, inline/inject would be selected in place of cuda IPC, and gdrcopy is used instead. So we are actually optimizing the **current** cuda IPC path(by not doing cuda ipc).

As a result, in such cases we know there is only 1 iov with `FI_HMEM_CUDA`, so we don't have to worry about mixed ifaces.

Does this help clear things up?

Reply:
Ah ok thank you that makes more sense :)
When you say "we know there is only 1 iov with `FI_HMEM_CUDA`" are you saying that's the case you're trying to target for the gdrcopy optimization? If there are multiple gdr vapable iovs in the array, is it less optimal to do iov_count number of gdr copies? I guess I'm trying to understand why we are only targeting the single iov rather than iterating over the array.

Reply:
> why we are only targeting the single iov rather than iterating over the array.

This is right on - yes this PR only addresses/improves the single-iov case(which has been my focus so far), while not touching the mixed iov path. In other words, this is the minimal impl to make it "work".

I also recognize the opportunity to make it work for mixed iovs. We could do what you suggested, change common mr iov copy code, but I prefer to do it in `smr_copy_from_mr_iov`, so that each provider, e.g. shm, can explicitly request gdrcopy.

> is it less optimal to do iov_count number of gdr copies

It should be equally performant. We will be iterating over the iov and call gdrcopy individually. Reference: https://github.com/ofiwg/libfabric/pull/8836

If it sounds good to you, I will go ahead and change `smr_copy_from_mr_iov` to support mixed iov, which also shortens the `if(...)` conditions.

Reply:
@wenduwan Thanks! Yeah I think we're on the same page. My concern is mostly about the number of checks it inserts into the normal path. Something generic (but also still in shm to make the gdr decision as you suggest) is best. I'll trust whatever you come up with!
Also I think we're ok not checking the fi version here. The ofi_mr is not coming from the user but the provider/utility code which will allocate the full MR and would just set the fields to 0. 
Also the OFI convention for compares is var == val for readability ie `iov_count == 1 && mr == NULL`

Reply:
@aingerson I stole code from [here](https://github.com/ofiwg/libfabric/blob/main/prov/efa/src/efa_hmem.h#L80) and [here](https://github.com/ofiwg/libfabric/blob/main/src/hmem.c#L271) in the update. This causes some duplication (passable IMO) so let me know if you think it's excessive.

Reply:
@wenduwan I see what you mean about the duplicate code... I'm leaning towards saying to just do it in the mr iov path with maybe a compile check around the gdr check? shm is the only provider using the mr iov copy functions and since the gdrcopy hmem data is in the ofi mr, I think it makes sense to say if you're going to use those functions that you would be opting in to the gdr copy if the gdr copy handle was registered it.
if there comes another provider/use case that needs to use it that explicitly needs to avoid the gdr code on a system with it available when it was registered with the gdr handle in the hmem data... then I think we can cross that bridge when we come to it.

Reply:
@aingerson Sneaky! Yeah if SHM is the only user then we can definitely do that.

But I just realized that there is another caveat. For gdrcopy to work, the caller should make sure the src and dest cannot both be on cuda. Based on my reading SHM does not have this issue(need to run more tests), so I will leave a note in code.

Reply:
Oh if that's the case, then we can never use gdr copy for the inline/inject paths since those always copy to the shm region which will never be cuda. If we have to make sure they are both cuda then we can only call it on the progress ipc path if the cmd->iface and rx_entry->iface are both cuda. I think it would be best to keep the ofi_mr copy functions as is and just add a case in the progress ipc path to call the gdrcopy function

Reply:
@aingerson lol I think you read this in reverse?

>  For gdrcopy to work, the caller should make sure the src and dest **cannot** both be on cuda

Reply:
@wenduwan oh lol yup. my bad!